File: SampleApps/Objective-C/PayUmoneySampleApp_ObjC/ViewController.m
//
//  ViewController.m
//  frameworkIntegrationFromScratch
//
//  Created by Vipin Aggarwal on 16/01/17.
//  Copyright © 2017 PayUmoney. All rights reserved.
//

#import "ViewController.h"
#import <PayUmoney_SDK/PayUmoney_SDK.h>
#import <CommonCrypto/CommonDigest.h>  //No need to import it in production app

@interface ViewController ()

@property (nonatomic, strong) PUMRequestParams *params;
@property (weak, nonatomic) IBOutlet UITextField *amount;
@property (weak, nonatomic) IBOutlet UITextField *firstname;
@property (weak, nonatomic) IBOutlet UITextField *email;
@property (weak, nonatomic) IBOutlet UITextField *productinfo;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}

- (IBAction)startPaymentTapped:(id)sender {
    
    [self setPaymentParameters];
    
    //Start the payment flow
    PUMMainVController *paymentVC = [[PUMMainVController alloc] init];
    UINavigationController *paymentNavController = [[UINavigationController alloc] initWithRootViewController:paymentVC];
    
    [self presentViewController:paymentNavController
                       animated:YES
                     completion:nil];
}

- (void)setPaymentParameters {
    self.params = [PUMRequestParams sharedParams];
    self.params.environment = PUMEnvironmentProduction;
    self.params.amount = self.amount.text;
    self.params.key = @"mdyCKV";
    self.params.merchantid = @"4914106";
    self.params.txnid = [self  getRandomString:2];
    self.params.surl = @"https://www.payumoney.com/mobileapp/payumoney/success.php";
    self.params.furl = @"https://www.payumoney.com/mobileapp/payumoney/failure.php";
    self.params.delegate = self;
    self.params.firstname = self.firstname.text;
    self.params.productinfo = self.productinfo.text;
    self.params.email = self.email.text;
    self.params.phone = @"";
    
    //Below parameters are optional. It is to store any information you would like to save in PayU Database regarding trasnsaction. If you do not intend to store any additional info, set below params as empty strings.
    
    self.params.udf1 = @"";
    self.params.udf2 = @"";
    self.params.udf3 = @"";
    self.params.udf4 = @"";
    self.params.udf5 = @"";
    self.params.udf6 = @"";
    self.params.udf7 = @"";
    self.params.udf8 = @"";
    self.params.udf9 = @"";
    self.params.udf10 = @"";
    
    self.params.hashValue = [self getHash];
    
}

- (NSString *)getRandomString:(NSInteger)length {
    NSMutableString *returnString = [NSMutableString stringWithCapacity:length];
    NSString *numbers = @"0123456789";
    
    // First number cannot be 0
    [returnString appendFormat:@"%C", [numbers characterAtIndex:(arc4random() % ([numbers length]-1))+1]];
    
    for (int i = 1; i < length; i++) {
        [returnString appendFormat:@"%C", [numbers characterAtIndex:arc4random() % [numbers length]]];
    }
    return returnString;
}


#pragma mark - Never Generate hash from app
/*!
    Keeping salt in the app is a big security vulnerability. Never do this. Following function is just for demonstratin purpose
    In code below, salt Je7q3652 is mentioned. Never do this in prod app. You should get the hash from your server.
 */

- (NSString*)getHash {
    NSString *hashSequence = [NSString stringWithFormat:@"mdyCKV|%@|%@|%@|%@|%@|||||||||||Je7q3652",self.params.txnid, self.params.amount, self.params.productinfo,self.params.firstname, self.params.email];
    
    NSString *rawHash = [[self createSHA512:hashSequence] description];
    NSString *hash = [[[rawHash stringByReplacingOccurrencesOfString:@"<" withString:@""]stringByReplacingOccurrencesOfString:@">" withString:@""]stringByReplacingOccurrencesOfString:@" " withString:@""];
    
    return hash;
}

- (NSData *) createSHA512:(NSString *)source {
    
    const char *s = [source cStringUsingEncoding:NSASCIIStringEncoding];
    NSData *keyData = [NSData dataWithBytes:s length:strlen(s)];
    uint8_t digest[CC_SHA512_DIGEST_LENGTH] = {0};
    CC_SHA512(keyData.bytes, (CC_LONG)keyData.length, digest);
    
    NSData *output = [NSData dataWithBytes:digest length:CC_SHA512_DIGEST_LENGTH];
    NSLog(@"out --------- %@",output);
    return output;
}



#pragma mark - Completeion callbacks

-(void)transactionCompletedWithResponse:(NSDictionary*)response
                       errorDescription:(NSError* )error {
  [self dismissViewControllerAnimated:YES completion:nil];
  [self showAlertViewWithTitle:@"Message" message:@"congrats! Payment is Successful"];
}

/*!
 * Transaction failure occured. Check Payment details in response. error shows any error
 if api failed.
 */
-(void)transactinFailedWithResponse:(NSDictionary* )response
                   errorDescription:(NSError* )error {
  [self dismissViewControllerAnimated:YES completion:nil];
  [self showAlertViewWithTitle:@"Message" message:@"Oops!!! Payment Failed"];
}

-(void)transactinExpiredWithResponse: (NSString *)msg {
  [self dismissViewControllerAnimated:YES completion:nil];
  [self showAlertViewWithTitle:@"Message" message:@"Trasaction expired!"];
}

/*!
 * Transaction cancelled by user.
 */
-(void)transactinCanceledByUser {
  [self dismissViewControllerAnimated:YES completion:nil];
  [self showAlertViewWithTitle:@"Message" message:@"Payment Cancelled!"];
}

#pragma mark - Helper methods

- (void)showAlertViewWithTitle:(NSString*)title message:(NSString*)message {
  UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title
                                                                           message:message preferredStyle:UIAlertControllerStyleAlert];
  
  UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
  [alertController addAction:ok];
  
  [self presentViewController:alertController animated:YES completion:nil];
}
@end

File: SampleApps/Swift/PayUmoneySampleApp_Swift/ViewController.swift
//
//  ViewController.swift
//  frameworkIntegration
//
//  Created by Honey Lakhani on 1/16/17.
//  Copyright © 2017 com.payu. All rights reserved.
//

import UIKit


class ViewController: UIViewController {
  //MARK: Stored properties
  var params : PUMRequestParams = PUMRequestParams.shared()
  var utils : Utils = Utils()
  
  //MARK: IBOutlets
  @IBOutlet weak var txtFldAmt: UITextField!
  @IBOutlet weak var txtFldName: UITextField!
  @IBOutlet weak var txtFldEmail: UITextField!
  @IBOutlet weak var txtFldProductInfo: UITextField!
  
  //MARK: IBActions
  @IBAction func btnSubmitClicked(_ sender: Any) {
    startPayment()
  }
  
  //MARK: UIView lifecycle methods
  override func viewDidLoad() {
    super.viewDidLoad()
  }
  
  //MARK: Helper methods
  func startPayment() -> Void {
    if((txtFldAmt.text) != nil){
      if (Double(txtFldAmt.text!)! > 1000000.00) {
        showAlertViewWithTitle(title: "Amount Exceeding the limit", message: "1000000")
        return
      }
      params.amount = txtFldAmt.text;
    }
     //PUMEnvironment.test for test environment and PUMEnvironment.production for live environment.
      params.environment = PUMEnvironment.test;
      params.firstname = txtFldName.text;
      params.key = "40747T";
      params.merchantid = "396132";  //Merchant merchantid
      params.logo_url = ""; //Merchant logo_url
      params.productinfo = "Product Info";
      params.email = txtFldEmail.text;  //user email
      params.phone = ""; //user phone
      params.txnid = utils.getRandomString(2);  //set your correct transaction id here
      params.surl = "https://www.payumoney.com/mobileapp/payumoney/success.php";
      params.furl = "https://www.payumoney.com/mobileapp/payumoney/failure.php";
    
    //Below parameters are optional. It is to store any information you would like to save in PayU Database regarding trasnsaction. If you do not intend to store any additional info, set below params as empty strings.
    
      params.udf1 = "";
      params.udf2 = "";
      params.udf3 = "";
      params.udf4 = "";
      params.udf5 = "";
      params.udf6 = "";
      params.udf7 = "";
      params.udf8 = "";
      params.udf9 = "";
      params.udf10 = "";
    //We strictly recommend that you calculate hash on your server end. Just so that you can quickly see demo app working, we are providing a means to do it here. Once again, this should be avoided.
    if(params.environment == PUMEnvironment.production){
      generateHashForProdAndNavigateToSDK()
    }
    else{
     calculateHashFromServer()
    }
      // assign delegate for payment callback.
    params.delegate = self;
     }
  
  func generateHashForProdAndNavigateToSDK() -> Void {
    let txnid = params.txnid!
    //add your salt in place of 'salt' if you want to test on live environment.
    //We suggest to calculate hash from server and not to keep the salt in app as it is a severe security vulnerability.
    let hashSequence : NSString = "\(params.key)|\(txnid)|\(params.amount)|\(params.productinfo)|\(params.firstname)|\(params.email)|||||||||||salt" as NSString
    let data :NSString = utils.createSHA512(hashSequence as String!) as NSString
    params.hashValue = data as String!;
       startPaymentFlow();
  }


  func transactinCanceledByUser() -> Void {
    self.dismiss(animated: true){
      self.showAlertViewWithTitle(title: "Message", message: "Payment Cancelled ")
    }
  }
 
  func startPaymentFlow() -> Void {
    let paymentVC : PUMMainVController = PUMMainVController()
    var paymentNavController : UINavigationController;
    paymentNavController = UINavigationController(rootViewController: paymentVC);
    self.present(paymentNavController, animated: true, completion: nil)
  }
  
  func transactionCompleted(withResponse response : NSDictionary,errorDescription error:NSError) -> Void {
    self.dismiss(animated: true){
      self.showAlertViewWithTitle(title: "Message", message: "congrats! Payment is Successful")
    }
  }
  

  func transactinFailed(withResponse response : NSDictionary,errorDescription error:NSError) -> Void {
    self.dismiss(animated: true){
      self.showAlertViewWithTitle(title: "Message", message: "Oops!!! Payment Failed")
    }
  }
  
  func showAlertViewWithTitle(title : String,message:String) -> Void {
    let alertController : UIAlertController = UIAlertController(title: title, message: message, preferredStyle: UIAlertControllerStyle.alert)
    let okAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.default) {
      UIAlertAction in
      NSLog("OK Pressed")
    }
    alertController.addAction(okAction)
    self.present(alertController, animated: true, completion: nil)
  }

// MARK:HASH CALCULATION
  
  //hash calculation strictly recommended to be done on your server end. This is just to show the hash sequence format and oe the api call for hash should be. Encryption is SHA-512.
 func prepareHashBody()->NSString{
  return "key=\(params.key!)&amount=\(params.amount!)&txnid=\(params.txnid!)&productinfo=\(params.productinfo!)&email=\(params.email!)&firstname=\(params.firstname!)" as NSString;
  }

 func calculateHashFromServer(){
  let config = URLSessionConfiguration.default // Session Configuration
  let session = URLSession(configuration: config) // Load configuration into Session
  let url = URL(string: "https://test.payumoney.com/payment/op/v1/calculateHashForTest")!
  var request = URLRequest(url: url)
  request.httpBody = prepareHashBody().data(using: String.Encoding.utf8.rawValue)
  request.httpMethod = "POST"
  
  let task = session.dataTask(with: request, completionHandler: {
    (data, response, error) in
    if error != nil {
      print(error!.localizedDescription)
    } else {
      do {
        if let json = try JSONSerialization.jsonObject(with: data!, options: []) as? [String: Any]{
          //Implement your logic
          print(json)
          let status : NSNumber = json["status"] as! NSNumber
          if(status.intValue == 0)
          {
           self.params.hashValue = json["result"] as! String!
            OperationQueue.main.addOperation {
              self.startPaymentFlow()
            }
          }
          else{
            OperationQueue.main.addOperation {
            self.showAlertViewWithTitle(title: "Message", message: json["message"] as! String)
            }
          }
        }
      } catch {
        print("error in JSONSerialization")
      }
    }
  })
    task.resume()
  }
}
